deployments/common/crds/k8s.nginx.org_virtualservers.yaml

```diff
@@ -479,6 +479,26 @@ spec:
                   description: TLS defines TLS configuration for a VirtualServer.
                   type: object
                   properties:
+                    cert-manager:
+                      description: CertManager defines a cert manager config for a TLS.
+                      type: object
+                      properties:
```

Comment:
Are any of these fields required? Can they all be empty string?

This will validate:
```
cert-manager: {}
```

But will it configure anything?

Reply:
Keeping with the pattern for the ingress and gateway controller shims, the validation happens within the cert-manager controller - see:

https://github.com/nginxinc/kubernetes-ingress/pull/2572/files/9a4f092c3121bd7fe591418830148a3426e38519#diff-5aceb73e31c7cb433bbd6cdb4664756a743795cea4714c2fa9773583dde5f8e7R257

and

https://github.com/nginxinc/kubernetes-ingress/pull/2572/files/9a4f092c3121bd7fe591418830148a3426e38519#diff-cd557a292871ac838e3b39960721fcb038ab8a639834fd70e404518d63a67549R71

None of the fields are required by themselves, but for example one or the other of Issuer and ClusterIssuer must be defined. Defining both gives the following error:

`I0406 10:10:19.694289       1 event.go:285] Event(v1.ObjectReference{Kind:"VirtualServer", Namespace:"default", Name:"cafe", UID:"cbadb574-dee7-4271-9d23-86045a7cee27", APIVersion:"k8s.nginx.org/v1", ResourceVersion:"42066", FieldPath:""}): type: 'Warning' reason: 'BadConfig' Could not determine issuer for virtual server due to bad config: both "tls.cert-manager.issuer" and "tls.cert-manager.cluster-issuer" may not be set`

Defining nothing (or everything as an empty string) gives:

I0406 10:13:05.598696       1 event.go:285] Event(v1.ObjectReference{Kind:"VirtualServer", Namespace:"default", Name:"cafe", UID:"cbadb574-dee7-4271-9d23-86045a7cee27", APIVersion:"k8s.nginx.org/v1", ResourceVersion:"42415", FieldPath:""}): type: 'Warning' reason: 'BadConfig' Could not determine issuer for virtual server due to bad config: failed to determine issuer name to be used for virtualserver resource

Reply:
I see, 'issuer' and 'cluster-issuer' are mutually exclusive. Exclusion is a pattern I think we can error closer to the user when they are setting the VirtualServer.

For instance:
```
kc apply -f lock-fail.yaml 
The Lock "lock-test" is invalid: <nil>: Invalid value: "": "spec" must validate one and only one schema (oneOf). Found 2 valid alternatives
```

You can do this with `oneOf`, then the case issuer and cluster-issuer are set, k8s API server rejects the VirtualServer and the user gets immediate feedback it's broken.

E.g.)
```
# CRD snippet:
    schema:
      openAPIV3Schema:
        properties:
          spec:
            type: object
            description: "test"
            properties:
              locked:
                type: string
              unlocked:
                type: string
            oneOf:
              - properties:
                  unlocked:
                    maximum: 0
                required: ["locked"]
              - properties:
                  locked:
                    maximum: 0
                required: ["unlocked"]
        type: object
---
# succeeds
apiVersion: lock.example.com/v1alpha1
kind: Lock
metadata:
  name: locked-test
spec:
  locked: "yes"
---
# succeeds
apiVersion: lock.example.com/v1alpha1
kind: Lock
metadata:
  name: unlocked-test
spec:
  unlocked: "yes"
---
# fails
apiVersion: lock.example.com/v1alpha1
kind: Lock
metadata:
  name: lock-test
spec:
  locked: "yes"
  unlocked: "yes"
```